custom_components/localtuya/coordinator.py

```diff
@@ -525,19 +532,34 @@ async def _async_reconnect(self):
                 continue
 
             try:
+                if not self._connect_task:
+                    await self.async_connect()
```

Comment:
No: the order is the fix for the case when `self.async_connect()` does not wait for `connect_task` to complete!

Reply:
What if the task for some reason is being canceled and has to be waited before call connect? 
I think this will works much better instead wait the connect task then ignore the connect with "else", Check if it's connected, If not call the connect.
```py
                if self._connect_task:
                    await self._connect_task

                if not self.connected:
                    await self.async_connect()
```

Reply:
> What if the task for some reason is being canceled and has to be waited before call connect?

`async_connect()` does a lot of checks before creating a task. In this case, it will do nothing. Then,
```
                if self._connect_task:
                    await self._connect_task
```
will wait for the cancelled task to complete. Then the following condition
https://github.com/xZetsubou/hass-localtuya/blob/a34e316448b5dd1762afee5b0c696016c5d756d9/custom_components/localtuya/coordinator.py#L542-L544
is `False`, and the loop will be repeated in the next `RECONNECT_INTERVAL` seconds to try to reconnect. Here the delay before reconnect attempt is not bad, because of the reason the task was cancelled. So, here my version should work correctly.

But the most usual case when we need just to re-connect. Note, that `async_connect()` starts with a lot of checks, and creates a new task only if it can be done. In most of cases, it should create the task. But it ends with
https://github.com/xZetsubou/hass-localtuya/blob/a34e316448b5dd1762afee5b0c696016c5d756d9/custom_components/localtuya/coordinator.py#L141-L142
and will not wait for the task to complete for sub-devices. Then, the check
https://github.com/xZetsubou/hass-localtuya/blob/a34e316448b5dd1762afee5b0c696016c5d756d9/custom_components/localtuya/coordinator.py#L542-L544
is `False`, and it waits `RECONNECT_INTERVAL` to repeat the re-connect, while the task is already being performed. This exactly leads to delayed and misleading log records "Reconnect succeeded on attempt:" that I [showed here](https://github.com/xZetsubou/hass-localtuya/pull/305#issuecomment-2227186659).

Please also note, that `_async_reconnect()` is started only on disconnect event.
- For WiFi devices, this is the only way to re-connect, so that the condition `if self._connect_task:` as the first check is always false.
- A Zigbee/BLE subdevice can be re-connected also by the (fake) gateway. But, again, then checks in the `async_connect()` would prevent creating a redundant task.

So, it should work correctly for both WiFi devices and Zigbee/BLE sub-devices, without excess loop cycles and misleading messages.